auth

dotnize opened the pull request:
wat i did (in the first commit):
- removed package-lock.json (to keep packages updated automatically, especially in heroku)
- removed `snowpack` package (just bcos i wasnt sure if we needed it, u can add it again if its needed lol)
- more config entries for oauth2
- added `passport`, `passport-discord`, `express-session` dependencies
- modular routes
- and the auth stuffffffff

notes:
- current behavior: user requests webplayer page > if no session, redirect to `/auth`; else just serve the webplayer page. we can add some sort of homepage if you want, with a "login with discord" button that redirects to auth
- i just used a global array for storing the user auth data. i was gonna use a database like mongodb but if we're hosting this on heroku, i dont think the database content will remain (and we don't really need to keep the data permanently) when the process restarts so yeah just an array since it's just temporary stuff. i'm not sure if i implemented this correctly either so idk ill try exploring more options soon (like [redis](https://www.npmjs.com/package/redis), since [Heroku supports this](https://devcenter.heroku.com/articles/heroku-redis))
- the session check stuff could probably use some improvements (`index.js line 27`)
- i kept the console log stuff to track the auth & session process, for debugging purposes
- ~~still gotta figure out a way to pass guilds data to the frontend (and display all guilds that **both** the discordbot and the user is a member of)~~ i added an API route for requesting session data. this can be called from the frontend via a HTTP POST request to `/session` and returns the user json which looks like this
![image](https://user-images.githubusercontent.com/48910077/154880275-9767bc90-496a-4c09-9ca5-8c4619ba1e94.png)
with this data we can verify which guilds the user has access to, then display them on the frontend if the discord bot is also in those guilds


ps, from http://expressjs.com/en/resources/middleware/session.html:
> Warning The default server-side session storage, MemoryStore, is purposely not designed for a production environment. It will leak memory under most conditions, does not scale past a single process, and is meant for debugging and developing.

index.js line 18, ill probably just deal with this later (or when we switch to a database)


ps:
make sure u have set the Redirect URI in your [bot application](https://discord.com/developers/applications) to this (change the port accordingly)
![image](https://user-images.githubusercontent.com/48910077/154857958-aab9fe26-5f4e-4d45-b294-b17f6593997f.png)

Commit:
auth stuff

Commit:
only store the needed session data

Commit:
API endpoint for requesting session data

Commit:
replace local guild storage and just fetch guilds from session & bot
- removed the localstorage stuff
- added global availableGuilds array and session object
- getGuilds and recGuilds events
- updateGuildSelect() only gets called after updating availableGuilds

JoshCunningHum commented:
I am completely speechless and grateful and I apologize for being silent for several days, got some ongoing commissions so I can't really work on this for now. I will do the optimization (if there's any) and might do the welcome page if I have the time.

For now, I'll just merge this commit and will try to implement some improvement (if I can find some) once I got back from my commissions.

P.S. I don't need the snowpack, forgot to remove it

P.P.S. On second thought, since you're saying it needs some improvement, then I'll just keep this PR for now. I haven't really tested it and would probably not have the time to do so in these coming days.

Commit:
use .filter() & .find() for updating available guilds
instead of 2 forEach loops
- totally forgot those 2 methods existed lol
- used !! to convert .find()'s return value to boolean
- some comments for ideas on what to do with session

dotnize commented:
that's completely fine hahaha. i'm also gonna be inactive with this for the next few days cos i gotta catch up with school stuff

JoshCunningHum committed:
Send static files if Guild_ID is found in the link

Commit:
just use global properties directly instead of using functions

JoshCunningHum committed:
Merge branch 'auth' of https://github.com/JoshCunningHum/Dodong-webplayer into auth

JoshCunningHum committed:
functional local restructure

 - Add the backend server to socket network
 - Functional player buttons

JoshCunningHum committed:
sets the requestor from session data

JoshCunningHum committed:
proper grammar xd
